## Goal:
- Read [x]
- Some tests + some features [x]


Note: we think that gender and fare matter for the survival (we haven't check it)

- export it to json []

### Reading the data

In [1]:
"""
Titanic Data Analysis and JSON Export
Author: [Your Name]
Description: Analyze Titanic passenger data, engineer features, and export to JSON
"""

import pandas as pd
import numpy as np
import json

from pathlib import Path

# Set up paths
DATA_DIR = Path("data")
CSV_FILE = DATA_DIR / "titanic.csv"
JSON_FILE = DATA_DIR / "titanic_data.json"

# Create data directory if it doesn't exist
DATA_DIR.mkdir(exist_ok=True)

print("Project setup complete!")
print(f"Data directory: {DATA_DIR}")
print(f"CSV file location: {CSV_FILE}")


Project setup complete!
Data directory: data
CSV file location: data/titanic.csv


In [2]:
JSON_FILE

PosixPath('data/titanic_data.json')

In [3]:
df = pd.read_csv(CSV_FILE)

In [4]:
print(f"Dataset loaded successfully! Shape: {df.shape}")
print(f"\nColumns: {list(df.columns)}")
print(f"\nFirst few rows:")
print(df.head())

Dataset loaded successfully! Shape: (891, 12)

Columns: ['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']

First few rows:
   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 175

In [5]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


### Calculating Statistics (Step 3)

In [6]:
# Select numeric columns only
numeric_columns = df.select_dtypes(include=[np.number])

# Calculate statistics
statistics = pd.DataFrame({
    'Mean':   numeric_columns.mean(),
    'Median': numeric_columns.median(),
    'Std Dev': numeric_columns.std()
})

print(statistics)

                   Mean    Median     Std Dev
PassengerId  446.000000  446.0000  257.353842
Survived       0.383838    0.0000    0.486592
Pclass         2.308642    3.0000    0.836071
Age           29.699118   28.0000   14.526497
SibSp          0.523008    0.0000    1.102743
Parch          0.381594    0.0000    0.806057
Fare          32.204208   14.4542   49.693429


In [7]:
statistics

,Mean,Median,Std Dev
PassengerId,446.000000,446.0000,257.353842
Survived,0.383838,0.0000,0.486592
Pclass,2.308642,3.0000,0.836071
Age,29.699118,28.0000,14.526497
SibSp,0.523008,0.0000,1.102743
Parch,0.381594,0.0000,0.806057
Fare,32.204208,14.4542,49.693429


### Step 4

In [8]:
print("\n" + "="*50)
print("MISSING VALUES ANALYSIS")
print("="*50)

missing_data = {}
for col in df.columns:
    missing_count = df[col].isna().sum()
    missing_percent = (missing_count / len(df)) * 100
    missing_data[col] = {
        "missing_count": missing_count,
        "missing_percent": missing_percent
    }
    
missing_df = pd.DataFrame(missing_data).T
missing_df.sort_values("missing_count", ascending=False)


MISSING VALUES ANALYSIS


,missing_count,missing_percent
Cabin,687.0,77.104377
Age,177.0,19.865320
Embarked,2.0,0.224467
PassengerId,0.0,0.000000
Name,0.0,0.000000
Pclass,0.0,0.000000
Survived,0.0,0.000000
Sex,0.0,0.000000
Parch,0.0,0.000000
SibSp,0.0,0.000000


In [9]:
missing_df

,missing_count,missing_percent
PassengerId,0.0,0.000000
Survived,0.0,0.000000
Pclass,0.0,0.000000
Name,0.0,0.000000
Sex,0.0,0.000000
Age,177.0,19.865320
SibSp,0.0,0.000000
Parch,0.0,0.000000
Ticket,0.0,0.000000
Fare,0.0,0.000000


### Step 5

- Create a "FamilySize" feature (SibSp + Parch + 1)
- Create an "IsAlone" feature (FamilySize == 1)
- Create an "AgeGroup" feature (categorize age into groups)
- Analyze how these features differ between survivors and non-survivors

In [10]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='str')

p1: fix what it is missing --> "your code here"

In [11]:
# Create a copy so we don't mess up the original data
df_features = df.copy()

# Feature 1: Family Size — person + siblings/spouse + parents/children
df_features['FamilySize'] = df_features['SibSp'] + df_features['Parch'] + 1
print("Family Size sample:")
print(df_features[['SibSp', 'Parch', 'FamilySize']].head(10))

# Feature 2: Is Alone — True if traveling alone
df_features['IsAlone'] = (df_features['FamilySize'] == 1).astype(int)
print("\nIs Alone sample:")
print(df_features[['FamilySize', 'IsAlone']].head(10))

# Feature 3: Age Groups
def categorize_age(age):
    if pd.isna(age):
        return 'Unknown'
    elif age < 18:
        return 'Child'
    elif age < 30:
        return 'YoungAdult'
    elif age < 50:
        return 'Adult'
    else:
        return 'Senior'

df_features['AgeGroup'] = df_features['Age'].apply(categorize_age)
print("\nAge Group sample:")
print(df_features[['Age', 'AgeGroup']].head(10))

Family Size sample:
   SibSp  Parch  FamilySize
0      1      0           2
1      1      0           2
2      0      0           1
3      1      0           2
4      0      0           1
5      0      0           1
6      0      0           1
7      3      1           5
8      0      2           3
9      1      0           2

Is Alone sample:
   FamilySize  IsAlone
0           2        0
1           2        0
2           1        1
3           2        0
4           1        1
5           1        1
6           1        1
7           5        0
8           3        0
9           2        0

Age Group sample:
    Age    AgeGroup
0  22.0  YoungAdult
1  38.0       Adult
2  26.0  YoungAdult
3  35.0       Adult
4  35.0       Adult
5   NaN     Unknown
6  54.0      Senior
7   2.0       Child
8  27.0  YoungAdult
9  14.0       Child


In [16]:
# Analyze feature differences between survivors and non-survivors
print("\n" + "="*50)
print("FEATURE ANALYSIS: SURVIVED vs NOT SURVIVED")
print("="*50)
 
# Here is an example to get you started:
print("\nFamily Size by Survival:")
family_survival = df_features.groupby('Survived')['FamilySize'].agg(['mean', 'median', 'std'])
print(family_survival)
 
# Statistical test: Do these features help differentiate?
print("\n" + "="*50)
print("FEATURE DIFFERENTIATION ANALYSIS")
print("="*50)
 
survived = df_features[df_features['Survived'] == 1]
not_survived = df_features[df_features['Survived'] == 0]
 
print("\nFamily Size:")
print(f"  Survived mean: {survived['FamilySize'].mean():.2f}")
print(f"  Not Survived mean: {not_survived['FamilySize'].mean():.2f}")
print(f"  Difference: {abs(survived['FamilySize'].mean() - not_survived['FamilySize'].mean()):.2f}")


FEATURE ANALYSIS: SURVIVED vs NOT SURVIVED

Family Size by Survival:
              mean  median       std
Survived                            
0         1.883424     1.0  1.830669
1         1.938596     2.0  1.186076

FEATURE DIFFERENTIATION ANALYSIS

Family Size:
  Survived mean: 1.94
  Not Survived mean: 1.88
  Difference: 0.06


In [18]:
df_features.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'FamilySize', 'IsAlone',
       'AgeGroup'],
      dtype='str')

In [21]:
# Analyze feature differences between survivors and non-survivors
print("\n" + "="*50)
print("FEATURE ANALYSIS: SURVIVED vs NOT SURVIVED")
print("="*50)
 
# Here is an example to get you started:
print("\nIs alone by Survival:")
family_survival = df_features.groupby('Survived')['IsAlone'].agg(['mean', 'median', 'std'])
print(family_survival)
 
# Statistical test: Do these features help differentiate?
print("\n" + "="*50)
print("FEATURE DIFFERENTIATION ANALYSIS")
print("="*50)
 
survived = df_features[df_features['Survived'] == 1]
not_survived = df_features[df_features['Survived'] == 0]
 
print("\n Is Alone:")
print(f"  Survived mean: {survived['IsAlone'].mean():.2f}")
print(f"  Not Survived mean: {not_survived['IsAlone'].mean():.2f}")
print(f"  Difference: {abs(survived['IsAlone'].mean() - not_survived['IsAlone'].mean()):.2f}")


FEATURE ANALYSIS: SURVIVED vs NOT SURVIVED

Is alone by Survival:
              mean  median       std
Survived                            
0         0.681239     1.0  0.466421
1         0.476608     0.0  0.500184

FEATURE DIFFERENTIATION ANALYSIS

 Is Alone:
  Survived mean: 0.48
  Not Survived mean: 0.68
  Difference: 0.20


p2: i have df as dataframe and ask it to calculate the steps without the starter code


"My goal is to:

- Create a "FamilySize" feature (SibSp + Parch + 1)
- Create an "IsAlone" feature (FamilySize == 1)
- Create an "AgeGroup" feature (categorize age into groups)
- Analyze how these features differ between survivors and non-survivors

I currently have a variable called df. 

Here is what it looks like when I do df.head()


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S


I am beginner in python and I want to understand the code step by step. 


In [15]:
## gunter
import pandas as pd

# 1. FamilySize = SibSp + Parch + 1
df["FamilySize"] = df["SibSp"] + df["Parch"] + 1

# 2. IsAlone = 1 if FamilySize is 1, otherwise 0
df["IsAlone"] = (df["FamilySize"] == 1).astype(int)

# 3. AgeGroup = categorize Age into groups
def categorize_age(age):
    if pd.isna(age):
        return "Unknown"
    elif age < 18:
        return "Child"
    elif age < 30:
        return "Young Adult"
    elif age < 50:
        return "Adult"
    else:
        return "Older Adult"

df["AgeGroup"] = df["Age"].apply(categorize_age)

# 4. Analyze differences between survivors and non survivors

print("Average survival by FamilySize")
print(df.groupby("FamilySize")["Survived"].mean())

print("\nAverage survival by IsAlone")
print(df.groupby("IsAlone")["Survived"].mean())

print("\nAverage survival by AgeGroup")
print(df.groupby("AgeGroup")["Survived"].mean())

print("\nCounts by AgeGroup and Survived")
print(pd.crosstab(df["AgeGroup"], df["Survived"]))

Average survival by FamilySize
FamilySize
1     0.303538
2     0.552795
3     0.578431
4     0.724138
5     0.200000
6     0.136364
7     0.333333
8     0.000000
11    0.000000
Name: Survived, dtype: float64

Average survival by IsAlone
IsAlone
0    0.505650
1    0.303538
Name: Survived, dtype: float64

Average survival by AgeGroup
AgeGroup
Adult          0.417969
Child          0.539823
Older Adult    0.364865
Unknown        0.293785
Young Adult    0.350554
Name: Survived, dtype: float64

Counts by AgeGroup and Survived
Survived       0    1
AgeGroup             
Adult        149  107
Child         52   61
Older Adult   47   27
Unknown      125   52
Young Adult  176   95


In [22]:
df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,FamilySize,IsAlone,AgeGroup
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,2,0,Young Adult
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,2,0,Adult
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,1,1,Young Adult
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S,2,0,Adult
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S,1,1,Adult
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S,1,1,Young Adult
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S,1,1,Young Adult
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S,4,0,Unknown
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C,1,1,Young Adult


In [13]:
print(df.head())

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  


## Step 6

In [23]:
JSON_FILE

PosixPath('data/titanic_data.json')

from the template

In [24]:
import json
import pandas as pd
import numpy as np
from datetime import datetime
from pathlib import Path

# ============================================================
# CONFIG
# ============================================================
JSON_FILE = Path('data/titanic_data.json')
JSON_FILE.parent.mkdir(parents=True, exist_ok=True)  # ✅ Creates 'data/' folder if missing


# ============================================================
# CLASS 1: Passenger
# ============================================================
class Passenger:
    """Represents a single passenger with all attributes."""

    def __init__(self, passenger_id, name, age, sex, survived, pclass,
                 fare, embarked=None, sibsp=None, parch=None,
                 family_size=None, is_alone=None, age_group=None, title=None):

        self.passenger_id = int(passenger_id)       if pd.notna(passenger_id) else None
        self.name         = str(name)               if pd.notna(name)         else None
        self.age          = round(float(age), 1)    if pd.notna(age)          else None
        self.sex          = str(sex)                if pd.notna(sex)          else None
        self.survived     = int(survived)           if pd.notna(survived)     else None
        self.pclass       = int(pclass)             if pd.notna(pclass)       else None
        self.fare         = round(float(fare), 2)   if pd.notna(fare)         else None
        self.embarked     = str(embarked)           if pd.notna(embarked)     else None
        self.sibsp        = int(sibsp)              if pd.notna(sibsp)        else None
        self.parch        = int(parch)              if pd.notna(parch)        else None
        self.family_size  = int(family_size)        if pd.notna(family_size)  else None
        self.is_alone     = int(is_alone)           if pd.notna(is_alone)     else None
        self.age_group    = str(age_group)          if pd.notna(age_group)    else None
        self.title        = str(title)              if pd.notna(title)        else None

    def to_dict(self):
        """Convert passenger to dictionary for JSON export."""
        return {
            'passenger_id': self.passenger_id,
            'name':         self.name,
            'title':        self.title,
            'age':          self.age,
            'age_group':    self.age_group,
            'sex':          self.sex,
            'survived':     self.survived,
            'pclass':       self.pclass,
            'fare':         self.fare,
            'embarked':     self.embarked,
            'sibsp':        self.sibsp,
            'parch':        self.parch,
            'family_size':  self.family_size,
            'is_alone':     self.is_alone,
        }


# ============================================================
# CLASS 2: TitanicDataset
# ============================================================
class TitanicDataset:
    """Encapsulates all data processing: cleaning, stats, feature engineering, export."""

    def __init__(self, dataframe):
        self.raw_df     = dataframe.copy()
        self.df         = dataframe.copy()
        self.passengers = []

        # Run all processing steps on init
        self._engineer_features()
        self._create_passengers()

    # ----------------------------------------------------------
    # STEP 1: Feature Engineering (from previous work)
    # ----------------------------------------------------------
    def _categorize_age(self, age):
        if pd.isna(age):    return 'Unknown'
        elif age < 18:      return 'Child'
        elif age < 30:      return 'Young Adult'
        elif age < 50:      return 'Adult'
        else:               return 'Senior'

    def _engineer_features(self):
        """Create FamilySize, IsAlone, AgeGroup features."""
        self.df['FamilySize'] = self.df['SibSp'] + self.df['Parch'] + 1
        self.df['IsAlone']    = (self.df['FamilySize'] == 1).astype(int)
        self.df['AgeGroup']   = self.df['Age'].apply(self._categorize_age)
        print("✅ Feature engineering complete (FamilySize, IsAlone, AgeGroup)")

    # ----------------------------------------------------------
    # STEP 2: Build Passenger objects
    # ----------------------------------------------------------
    def _create_passengers(self):
        """Instantiate a Passenger object for every row."""
        for idx, row in self.df.iterrows():
            p = Passenger(
                passenger_id = row.get('PassengerId', idx),
                name         = row.get('Name',        None),
                age          = row.get('Age',         None),
                sex          = row.get('Sex',         None),
                survived     = row.get('Survived',    None),
                pclass       = row.get('Pclass',      None),
                fare         = row.get('Fare',        None),
                embarked     = row.get('Embarked',    None),
                sibsp        = row.get('SibSp',       None),
                parch        = row.get('Parch',       None),
                family_size  = row.get('FamilySize',  None),
                is_alone     = row.get('IsAlone',     None),
                age_group    = row.get('AgeGroup',    None),
                title        = row.get('Title',       None),
            )
            self.passengers.append(p)
        print(f"✅ Created {len(self.passengers)} Passenger objects")

    # ----------------------------------------------------------
    # STEP 3: Missing Values Analysis (from previous work)
    # ----------------------------------------------------------
    def get_missing_values(self):
        """Return missing value counts and percentages per column."""
        missing_count   = self.raw_df.isnull().sum()
        missing_percent = (missing_count / len(self.raw_df) * 100).round(2)
        missing_df = pd.DataFrame({
            'missing_count':   missing_count,
            'missing_percent': missing_percent
        })
        return missing_df[missing_df['missing_count'] > 0].sort_values('missing_count', ascending=False).to_dict()

    # ----------------------------------------------------------
    # STEP 4: Statistics (from previous work)
    # ----------------------------------------------------------
    def get_statistics(self):
        """Return mean, median, std for all numeric columns."""
        numeric_df = self.df.select_dtypes(include='number')
        return {
            'mean':   numeric_df.mean().round(2).to_dict(),
            'median': numeric_df.median().round(2).to_dict(),
            'std':    numeric_df.std().round(2).to_dict(),
        }

    # ----------------------------------------------------------
    # STEP 5: Summary Statistics
    # ----------------------------------------------------------
    def get_summary_stats(self):
        """High-level survival and demographic statistics."""
        survived     = sum(1 for p in self.passengers if p.survived == 1)
        not_survived = sum(1 for p in self.passengers if p.survived == 0)
        ages         = [p.age  for p in self.passengers if p.age  is not None]
        fares        = [p.fare for p in self.passengers if p.fare is not None]

        # Survival by feature (from feature engineering analysis)
        survived_df     = self.df[self.df['Survived'] == 1]
        not_survived_df = self.df[self.df['Survived'] == 0]

        return {
            'total_passengers':       len(self.passengers),
            'survived':               survived,
            'did_not_survive':        not_survived,
            'survival_rate_%':        round(survived / len(self.passengers) * 100, 2),
            'average_age':            round(sum(ages)  / len(ages),  2) if ages  else None,
            'average_fare':           round(sum(fares) / len(fares), 2) if fares else None,
            'avg_family_size_survived':     round(survived_df['FamilySize'].mean(), 2),
            'avg_family_size_not_survived': round(not_survived_df['FamilySize'].mean(), 2),
            'alone_rate_survived_%':        round(survived_df['IsAlone'].mean() * 100, 2),
            'alone_rate_not_survived_%':    round(not_survived_df['IsAlone'].mean() * 100, 2),
        }

    # ----------------------------------------------------------
    # STEP 6: Export to JSON
    # ----------------------------------------------------------
    def to_json(self, filename=JSON_FILE):
        """Export everything — metadata, stats, missing values, passengers — to JSON."""
        filename = Path(filename)
        filename.parent.mkdir(parents=True, exist_ok=True)  # ensure folder exists

        data = {
            'metadata': {
                'dataset_name':     'Titanic Passenger Dataset',
                'export_date':      datetime.now().isoformat(),
                'total_passengers': len(self.passengers),
                'survival_rate_%':  round(self.df['Survived'].mean() * 100, 2),
                'source':           'Kaggle Titanic Dataset',
            },
            'summary_statistics':  self.get_summary_stats(),
            'descriptive_statistics': self.get_statistics(),
            'missing_values':      self.get_missing_values(),
            'passengers':          [p.to_dict() for p in self.passengers],
        }

        with open(filename, 'w') as f:
            json.dump(data, f, indent=2)

        print(f"✅ JSON exported to: {filename.resolve()}")
        return data


# ============================================================
# RUN
# ============================================================
if 'df' in locals() and not df.empty:
    dataset = TitanicDataset(df)

    print("\n" + "="*50)
    print("       TITANIC DATASET SUMMARY")
    print("="*50)
    for key, value in dataset.get_summary_stats().items():
        print(f"  {key.replace('_', ' ').title():<40} {value}")
    print("="*50)

    # Export to JSON
    dataset.to_json(JSON_FILE)

✅ Feature engineering complete (FamilySize, IsAlone, AgeGroup)
✅ Created 891 Passenger objects

       TITANIC DATASET SUMMARY
  Total Passengers                         891
  Survived                                 342
  Did Not Survive                          549
  Survival Rate %                          38.38
  Average Age                              29.7
  Average Fare                             32.2
  Avg Family Size Survived                 1.94
  Avg Family Size Not Survived             1.88
  Alone Rate Survived %                    47.66
  Alone Rate Not Survived %                68.12
✅ JSON exported to: /home/bella/Desktop/lab2/data/titanic_data.json


p2

In [30]:
# Export to JSON
df_features.to_json('data/titanic_data2.json')

# Validate — load it back and check the structure
with open('data/titanic_data2.json', 'r', encoding='utf-8') as f:
    json_data = json.load(f)

print("JSON Validation:")
print(f"  Keys in file: {list(json_data.keys())}")
print(f"\nFirst passenger in JSON:")


JSON Validation:
  Keys in file: ['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'FamilySize', 'IsAlone', 'AgeGroup']

First passenger in JSON:


### Step 7

In [31]:
# Additional validation: Load and inspect JSON
with open(JSON_FILE, 'r', encoding='utf-8') as f:
    json_data = json.load(f)

# Print summary of JSON data and verify content


In [32]:
json_data

{'metadata': {'dataset_name': 'Titanic Passenger Dataset',
  'export_date': '2026-04-22T12:12:20.855976',
  'total_passengers': 891,
  'survival_rate_%': 38.38,
  'source': 'Kaggle Titanic Dataset'},
 'summary_statistics': {'total_passengers': 891,
  'survived': 342,
  'did_not_survive': 549,
  'survival_rate_%': 38.38,
  'average_age': 29.7,
  'average_fare': 32.2,
  'avg_family_size_survived': 1.94,
  'avg_family_size_not_survived': 1.88,
  'alone_rate_survived_%': 47.66,
  'alone_rate_not_survived_%': 68.12},
 'descriptive_statistics': {'mean': {'PassengerId': 446.0,
   'Survived': 0.38,
   'Pclass': 2.31,
   'Age': 29.7,
   'SibSp': 0.52,
   'Parch': 0.38,
   'Fare': 32.2,
   'FamilySize': 1.9,
   'IsAlone': 0.6},
  'median': {'PassengerId': 446.0,
   'Survived': 0.0,
   'Pclass': 3.0,
   'Age': 28.0,
   'SibSp': 0.0,
   'Parch': 0.0,
   'Fare': 14.45,
   'FamilySize': 1.0,
   'IsAlone': 1.0},
  'std': {'PassengerId': 257.35,
   'Survived': 0.49,
   'Pclass': 0.84,
   'Age': 14.53,

In [36]:
json_data.keys()

dict_keys(['metadata', 'summary_statistics', 'descriptive_statistics', 'missing_values', 'passengers'])

In [33]:
type(json_data)

dict

In [35]:
json_data["summary_statistics"]

{'total_passengers': 891,
 'survived': 342,
 'did_not_survive': 549,
 'survival_rate_%': 38.38,
 'average_age': 29.7,
 'average_fare': 32.2,
 'avg_family_size_survived': 1.94,
 'avg_family_size_not_survived': 1.88,
 'alone_rate_survived_%': 47.66,
 'alone_rate_not_survived_%': 68.12}